libavfilter/vf_gblur_vulkan.c

```diff
@@ -0,0 +1,472 @@
+                .oldLayout                   = out->layout[i],
+                .newLayout                   = s->output_images[i].imageLayout,
+                .srcQueueFamilyIndex         = VK_QUEUE_FAMILY_IGNORED,
+                .dstQueueFamilyIndex         = VK_QUEUE_FAMILY_IGNORED,
+                .image                       = out->img[i],
+                .subresourceRange.aspectMask = VK_IMAGE_ASPECT_COLOR_BIT,
+                .subresourceRange.levelCount = 1,
+                .subresourceRange.layerCount = 1,
+            },
+        };
+
+        vkCmdPipelineBarrier(cmd_buf, VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT,
+                             VK_PIPELINE_STAGE_COMPUTE_SHADER_BIT, 0,
+                             0, NULL, 0, NULL, FF_ARRAY_ELEMS(barriers), barriers);
+
+        in->layout[i]  = barriers[0].newLayout;
+        in->access[i]  = barriers[0].dstAccessMask;
+
+        tmp->layout[i] = barriers[1].newLayout;
+        tmp->access[i] = barriers[1].dstAccessMask;
+
+        out->layout[i] = barriers[2].newLayout;
+        out->access[i] = barriers[2].dstAccessMask;
+    }
+
+    ff_vk_bind_pipeline_exec(avctx, s->exec, s->pl_hor);
+
+    vkCmdDispatch(cmd_buf, FFALIGN(s->vkctx.output_width, CGS)/CGS,
+                  FFALIGN(s->vkctx.output_height, CGS)/CGS, 1);
+
+    ff_vk_bind_pipeline_exec(avctx, s->exec, s->pl_ver);
+
+    vkCmdDispatch(cmd_buf, FFALIGN(s->vkctx.output_width, CGS)/CGS,
+                  FFALIGN(s->vkctx.output_height, CGS)/CGS, 1);
+
+    ff_vk_add_exec_dep(avctx, s->exec, inframe, VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT, 1);
+    ff_vk_add_exec_dep(avctx, s->exec, outframe, VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT, 0);
+
+    return ff_vk_submit_exec_queue(avctx, s->exec);
+
+fail:
+    ff_vk_discard_exec_deps(avctx, s->exec);
+    return err;
+}
+
+static int gblur_vulkan_filter_frame(AVFilterLink *link, AVFrame *in)
+{
+    int err;
+    AVFrame *out = NULL;
+    AVFrame *tmp = NULL;
+    AVFilterContext *ctx = link->dst;
+    GBlurVulkanContext *s = ctx->priv;
+    AVFilterLink *outlink = ctx->outputs[0];
+
+    out = ff_get_video_buffer(outlink, outlink->w, outlink->h);
+    if (!out) {
+        err = AVERROR(ENOMEM);
+        goto fail;
+    }
+
+    tmp = ff_get_video_buffer(outlink, outlink->w, outlink->h);
+    if (!tmp) {
+        err = AVERROR(ENOMEM);
+        goto fail;
+    }
+
+    if (!s->initialized)
+        RET(init_filter(ctx, in));
+
+    RET(process_frames(ctx, out, tmp, in));
+
+    RET(av_frame_copy_props(out, in));
+
+    av_frame_free(&in);
+    av_frame_free(&tmp);
+
+    return ff_filter_frame(outlink, out);
+
+fail:
+    av_frame_free(&in);
+    av_frame_free(&tmp);
+    av_frame_free(&out);
+
+    return err;
+}
+
+#define OFFSET(x) offsetof(GBlurVulkanContext, x)
+#define FLAGS (AV_OPT_FLAG_FILTERING_PARAM | AV_OPT_FLAG_VIDEO_PARAM)
+static const AVOption gblur_vulkan_options[] = {
+    { "sigma", "Set sigma", OFFSET(sigma), AV_OPT_TYPE_FLOAT, {.dbl = 0.5}, 0.001, 1024.0, FLAGS },
```

Comment:
why it's 0.001 not 0?

Reply:
The reason here is that using zero as the variance does nothing. As for 0.001, it's to avoid the user from entering a too-small value. Otherwise, the subsequent operation's a waste of resources and time.

Reply:
If we use 0, it will fail on GPU and success on CPU.